backend/app/tests/api/routes/inquiry/test_create_inquiry.py

```diff
@@ -0,0 +1,72 @@
+from unittest.mock import ANY, MagicMock, Mock
+
+import pytest
+from fastapi.testclient import TestClient
+from pytest import MonkeyPatch
+
+import app.services.inquiries
+from app.core.config import settings
+from app.models import Inquiry, InquiryCreate
+
+inquiry_text = "Why do birds suddenly appear every time you are near?"
+
+
+@pytest.fixture(autouse=True)
+def create_inquiry_mock(monkeypatch: MonkeyPatch) -> MagicMock:
+    _mock = MagicMock()
+    monkeypatch.setattr(
+        app.services.inquiries, app.services.inquiries.create_inquiry.__name__, _mock
+    )
+    return _mock
+
+
+@pytest.fixture(autouse=True)
+def get_inquiry_by_text_mock(monkeypatch: MonkeyPatch) -> MagicMock:
+    _mock = MagicMock()
+    monkeypatch.setattr(
+        app.services.inquiries,
+        app.services.inquiries.get_inquiry_by_text.__name__,
+        _mock,
+    )
+    return _mock
+
+
+def test_create_inquiry_api_calls_create_inquiry_service(
```

Comment:
here's an example where a unit test is testing more than one thing, and testing to implementation. Remember, the format should be:

createInquiry_whenCalledWithValidText_shouldCreateInquiryServiceOnce
createInquiry_whenCalledWithValidText_shouldCreateInquiryServiceWithPassedInText

Maybe something like that? (I'm not completely against testing both once and text, i.e. a UT doesn't only have to have a single assert, but generally when possible, it makes for less brittle tests. Especially if you factor out all the setup into a separate function)

Reply:
This test is testing one thing.  the api layer calls the service layer.  it requests that the service layer make a call to the database to create the Inquiry entity. the test uses a Spy to make sure that the function is invoked with the correct argument. I'm not sure what you are asking for here.

Reply:
The API layer calls the service layer
That 
(1) checks whether the Inquiry exists
get_inquiry_by_text_mock.assert_called_once()
(2) It doesn't so it asks the service layer to create the inquiry
    create_inquiry_mock.assert_called_once()
 (3) The service layer invokes its create_inquiry with the correct argument
    create_inquiry_mock.assert_called_with(
        session=ANY, inquiry_in=InquiryCreate(text=inquiry_text)
    )

Reply:
https://danywalls.com/how-and-when-to-use-mocks-or-spies-on-unit-testing#heading-spy

Reply:
this is the code being tested

python```
@router.post("/", response_model=InquiryPublic)
def create_inquiry(*, session: SessionDep, inquiry_in: InquiryCreate) -> Inquiry:
    """
    Create new inquiry.
    """
    inquiry = inquiries_service.get_inquiry_by_text(
        session=session, text=inquiry_in.text
    )
    if inquiry:
        raise HTTPException(
            status_code=400,
            detail="This inquiry already exists.",
        )

    return inquiries_service.create_inquiry(session=session, inquiry_in=inquiry_in)
```

Reply:
```
this is the code being tested

python```
@router.post("/", response_model=InquiryPublic)
def create_inquiry(*, session: SessionDep, inquiry_in: InquiryCreate) -> Inquiry:
    """
    Create new inquiry.
    """
    inquiry = inquiries_service.get_inquiry_by_text(
        session=session, text=inquiry_in.text
    )
    if inquiry:
        raise HTTPException(
            status_code=400,
            detail="This inquiry already exists.",
        )

    return inquiries_service.create_inquiry(session=session, inquiry_in=inquiry_in)
```


```

Reply:
Discussing offline.